### PR TITLE
Initial (partially-reviewed API) System.Net.Connections.

### PR DESCRIPTION
**Pull request by scalablecory**
This implements the already reviewed APIs of #1793. Followup PR will implement the Socket factories, otherwise this should match reviewed APIs.

Adds:
- System.Net.Connections assembly.

Makes these changes to existing code:
- System.IO.Pipelines is now included in Net50 SDK.
- SocketsHttpHandler properties added to connect via a connection factory rather than directly on a Socket.
- Test code: Loopback connections allows use without their accept servers.
- Test code: Moved settings read/write code from http/2 accept server into connection.
- Test code: VirtualNetwork supports read/write shutdown.

TODO:
- Implement tests for ConnectionExtensions and factories.
- Some additional tests for SocketsHttpHandler would be useful.

### CODE REVIEW

**Comment by scalablecory** [comment]
CC @JamesNK

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 1]
```diff
@@ -544,18 +545,28 @@ public byte[] Http2AltSvcOriginUri
...
                         await ConnectAsync(request, async, true, cancellationToken).ConfigureAwait(false);
+
                     if (failureResponse != null)
                     {
                         return (null, true, failureResponse);
                     }
 
+                    Debug.Assert(connection != null);
+
+                    sslStream = (SslStream)connection.Stream;
```
Comment (JamesNK) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:560]: I haven't looked at the context around this line closely, but have you tested your new logic with HTTP/2 + H2C? I'm guessing the stream won't be SslStream in that case.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Connections/SocketConnection.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            {
+                _stream.DisposeWithoutClosingConnection();
+            }
+
+            _socket.Dispose();
+        }
+
+        protected override Stream CreateStream()
+        {
+            return _stream = new SocketConnectionNetworkStream(this);
```
[the remark this answers is not in the document]
Comment (scalablecory) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Connections/SocketConnection.cs:59]: They're only called once, by the `Stream`/`Pipe` properties.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs (hunk 1 above)
Comment (scalablecory) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:560]: Tests pass but I don't recall what tests we have for H2C. I'll double check that this is correct, thanks.

File: src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        private IDuplexPipe? _pipe;
+        private bool _initializing;
+
+        /// <summary>
+        /// The connection's <see cref="Stream"/>.
+        /// </summary>
+        public Stream Stream =>
+            _stream != null ? _stream :
+            _pipe != null ? throw new InvalidOperationException("Unable to create Stream; Pipe already initialized.") :
+            (_stream = CreateStream() ?? throw new InvalidOperationException("CreateStream returned null."));
```
Comment (wfurt) [inline src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs:30]: should this return null instead of throwing?

File: src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+        /// At least one of <see cref="CreateStream"/> and <see cref="CreatePipe"/> must be overridden.
+        /// If only <see cref="CreateStream"/> is overridden, a user accessing <see cref="Pipe"/> will get a <see cref="IDuplexPipe"/> wrapping the <see cref="Stream"/>.
+        /// </remarks>
+        protected virtual Stream CreateStream()
+        {
+            if (_initializing) throw new InvalidOperationException("A Connection must implement one of CreateStream or CreatePipe.");
+
+            try
+            {
+                _initializing = true;
```
Comment (wfurt) [inline src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs:54]: This is not thread safe, right?

File: src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs (hunk 3 above)
Comment (wfurt) [inline src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs:30]: should this return null instead of throwing?

File: src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+            public override EndPoint? RemoteEndPoint { get; }
+
+            public ConnectionFromPipe(IDuplexPipe pipe, object? disposable, IConnectionProperties? properties, EndPoint? localEndPoint, EndPoint? remoteEndPoint)
+            {
+                Debug.Assert(disposable == null || disposable is IDisposable || disposable is IAsyncDisposable);
+
+                _pipe = pipe;
+                _disposable = disposable;
+                _properties = properties;
+                LocalEndPoint = localEndPoint;
```
Comment (wfurt) [inline src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs:232]: Why do we need this? For logging?

File: src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionBase.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        /// <summary>
+        /// Disposes of the connection.
+        /// </summary>
+        /// <remarks>This is equivalent to calling <see cref="CloseAsync(ConnectionCloseMethod, CancellationToken)"/> with the method <see cref="ConnectionCloseMethod.Immediate"/>.</remarks>
+        public void Dispose()
+        {
+            ValueTask t = CloseAsync(ConnectionCloseMethod.GracefulShutdown, CancellationToken.None);
+
+            if (t.IsCompleted) t.GetAwaiter().GetResult();
+            else t.AsTask().GetAwaiter().GetResult();
```
Comment (wfurt) [inline src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionBase.cs:65]: Do we need the AsTask?

File: src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+            }
+        }
+
+        private sealed class StreamFilteringFactory : ConnectionFactory
+        {
+            private readonly ConnectionFactory _baseFactory;
+            private readonly Func<Connection, IConnectionProperties?, CancellationToken, ValueTask<Stream>> _filter;
+
+            public StreamFilteringFactory(ConnectionFactory baseFactory, Func<Connection, IConnectionProperties?, CancellationToken, ValueTask<Stream>> filter)
+            {
```
Comment (wfurt) [inline src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs:108]: Why does it take CancellationToken if it is not used?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 8]
```diff
@@ -640,7 +650,7 @@ public byte[] Http2AltSvcOriginUri
                         Trace("Discarding downgraded HTTP/1.1 connection because connection limit is exceeded");
                     }
 
-                    sslStream.Close();
+                    await connection!.CloseAsync(ConnectionCloseMethod.Immediate, cancellationToken).ConfigureAwait(false);
```
Comment (wfurt) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:653]: Should this be graceful shutdown? We should send TLS alert IMHO.

**Review by wfurt — APPROVED** [review]
LGTM with few comments. 
But I'm not the expert on this.

**Comment by wfurt** [comment]
cc: @safern for any of the build changes.

File: src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs [hunk 9]
```diff
@@ -0,0 +1,288 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
```
Comment (safern) [inline src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs:3]: We no longer use this line for the license header. We just use the first two:


https://github.com/dotnet/runtime/commit/6072e4d3a7a2a1493f514cdf4be75a3d56580e84

**Review by safern — COMMENTED** [review]
I'll take a look at the build changes

File: src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs (hunk 9 above)
Comment (ManickaP) [inline src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs:3]: Old license header, the new has only the first two.

File: src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj [hunk 10]
```diff
@@ -8,278 +8,176 @@
...
-             Link="Common\System\Net\Http\HttpClientHandlerTestBase.cs" />
-    <Compile Include="$(CommonTestPath)System\Net\Http\TestHelper.cs"
-             Link="Common\System\Net\Http\TestHelper.cs" />
-    <Compile Include="$(CommonTestPath)System\Threading\TrackingSynchronizationContext.cs"
-             Link="Common\System\Threading\TrackingSynchronizationContext.cs" />
-    <Compile Include="$(CommonTestPath)System\Threading\Tasks\GetStateMachineData.cs"
-             Link="Common\System\Threading\Tasks\GetStateMachineData.cs" />
-    <Compile Include="$(CommonTestPath)System\Threading\Tasks\TaskTimeoutExtensions.cs"
-             Link="Common\System\Threading\Tasks\TaskTimeoutExtensions.cs" />
+    <Compile Include="$(CommonPath)Interop\Unix\Interop.Libraries.cs" Condition="'$(TargetsUnix)' == 'true' or '$(TargetsBrowser)' == 'true'" Link="Common\Interop\Unix\Interop.Libraries.cs" />
```
Comment (ManickaP) [inline src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj:11]: Please revert the unrelated changes made by VS. I.e.: `Link` attribute should be on the next line.

File: src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs (hunk 3 above)
Comment (scalablecory) [reply src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs:30]: Nope; CreateStream must return non-null.

File: src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs (hunk 5 above)
Comment (scalablecory) [reply src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs:232]: Yea. This implementation is for convenience, normal implementations will get local endpoint from e.g. a socket.

File: src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionBase.cs (hunk 6 above)
Comment (scalablecory) [reply src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionBase.cs:65]: `ValueTask` does not support synchronously waiting, so we do need a `Task`.

File: src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs (hunk 7 above)
Comment (scalablecory) [reply src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs:108]: It takes a `Func` that takes a `CancellationToken`.

File: src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+        /// <returns>A new <see cref="Connection"/>.</returns>
+        public static Connection FromPipe(IDuplexPipe pipe, IAsyncDisposable disposable, IConnectionProperties? properties = null, EndPoint? localEndPoint = null, EndPoint? remoteEndPoint = null)
+        {
+            if (pipe == null) throw new ArgumentNullException(nameof(pipe));
+            return new ConnectionFromPipe(pipe, disposable, properties, localEndPoint, remoteEndPoint);
+        }
+
+        private sealed class ConnectionFromStream : Connection, IConnectionProperties
+        {
+            private new Stream? _stream;
```
[the remark this answers is not in the document]
Comment (scalablecory) [reply src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs]: Can't use the Connection._stream as you might have this happen --

```c#
_pipe = CreatePipe() -> default implementation wraps CreateStream().
```

So I've renamed to `_originalStream` and `_originalPipe`.

Honestly it's weird that it is considered a shadowing when the shadowed variable is private and not so visible to that code :/

File: src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs (hunk 4 above)
Comment (scalablecory) [reply src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs]: Correct. None of this code is.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs (hunk 8 above)
Comment (scalablecory) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:653]: I don't know enough about TLS; I'll make it graceful.

File: src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj (hunk 10 above)
Comment (scalablecory) [reply src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj]: It's concerning that VS is doing this automatically. @safern do you know what is causing this to happen now? I don't want to revert this code with every new PR that alters project files :/

**Comment by scalablecory** [comment]
@safern can you please verify that the changes to System.IO.Pipelines are correct? Goal is to have this build as part of non-aspnet SDKs as System.Net.Http now depends on it.
@safern also that there's nothing else I need to do to add System.Net.Connections to SDKs.

File: src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj (hunk 10 above)
Comment (safern) [reply src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj]: IIRC, When you update something via VS or add a new file it trims the new lines. This was one of the downsides of moving the Link metadata declaration inline instead of using the extended XML tag format.

@davkean do you know how can we avoid this?

**Comment by safern** [comment]
Yeah, I'll look at this in the morning as these kind of changes need file navigation and in the phone is not ideal. Count on my review tomorrow morning 😊

File: src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj (hunk 10 above)
Comment (ManickaP) [reply src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj]: I talked about it with @ViktorHofer and he suggested to report it in VS. So I did, here's the [ticket](https://developercommunity.visualstudio.com/content/problem/1052431/csproj-file-gets-reformatted-upon-save.html). Please upvote if you want some reaction. Although, I'm highly skeptical about it.

File: src/libraries/System.IO.Pipelines/Directory.Build.props [hunk 12]
```diff
@@ -2,5 +2,6 @@
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
     <StrongNameKeyId>Open</StrongNameKeyId>
+    <IsNETCoreApp>true</IsNETCoreApp>
```
Comment (safern) [inline src/libraries/System.IO.Pipelines/Directory.Build.props:5]: I understand why we're making this inbox, but was people aware that we needed to make this inbox to support `System.Net.Connections`? 

You also need to update https://github.com/dotnet/runtime/blob/a366d638d880823ad63d28c22ddee0aeafbbcfc1/src/libraries/pkg/baseline/packageIndex.json#L3347-L3356

To add `"net5.0"` to `"InboxOn"` object.

File: src/libraries/System.Net.Connections/src/System.Net.Connections.csproj [hunk 13]
```diff
@@ -0,0 +1,28 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Library</OutputType>
```
Comment (safern) [inline src/libraries/System.Net.Connections/src/System.Net.Connections.csproj:3]: `OutputType` is not needed, we set it to `Library` in a common place.

File: src/libraries/System.Net.Connections/src/System.Net.Connections.csproj [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    <AssemblyName>System.Net.Connections</AssemblyName>
```
Comment (safern) [inline src/libraries/System.Net.Connections/src/System.Net.Connections.csproj:4]: `AssemblyName` property is not needed.

**Comment by ViktorHofer** [comment]
Note to myself, this will conflict with https://github.com/dotnet/runtime/pull/35606. If you are in first I need to react, otherwise you by declaring that the library is part of the shared framework in a different file.

File: src/libraries/System.IO.Pipelines/Directory.Build.props (hunk 12 above)
Comment (scalablecory) [reply src/libraries/System.IO.Pipelines/Directory.Build.props:5]: > I understand why we're making this inbox, but was people aware that we needed to make this inbox to support `System.Net.Connections`?

Yep, discussed in and out of API review

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs (hunk 1 above)
Comment (scalablecory) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: It seems we have no tests for H2C :). Thanks for pointing this out, fixed.

We'll want to add some tests for H2C, perhaps as part of https://github.com/dotnet/runtime/pull/39201 or elsewhere. I filed [an issue](https://github.com/dotnet/runtime/issues/39648) for it.

Comment (wfurt) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: H2C has some coverage on Mac where server-side ALPN is not available. So I fixed some tests to fall-back to HTTP/2 even without SSL.

File: src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+            }
+
+            protected override Stream CreateStream()
+            {
+                return _stream;
+            }
+
+            protected override async ValueTask CloseAsyncCore(ConnectionCloseMethod method, CancellationToken cancellationToken)
+            {
+                if (method != ConnectionCloseMethod.Immediate)
```
Comment (geoffkizer) [inline src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs:255]: Why do we flush in the abort case?

File: src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+    public static class ConnectionExtensions
+    {
+        /// <summary>
+        /// Retrieves a Type-based property from an <see cref="IConnectionProperties"/>, if it exists.
+        /// </summary>
+        /// <typeparam name="T">The type of the property to retrieve.</typeparam>
+        /// <param name="properties">The connection properties to retrieve a property from.</param>
+        /// <param name="property">If <paramref name="properties"/> contains a property of type <typeparamref name="T"/>, receives the property. Otherwise, default.</param>
+        /// <returns>If <paramref name="properties"/> contains a property of type <typeparamref name="T"/>, true. Otherwise, false.</returns>
+        public static bool TryGet<T>(this IConnectionProperties properties, [MaybeNullWhen(false)] out T property)
```
Comment (geoffkizer) [inline src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs:23]: I notice some TryGet variations use "MaybeNullWhen(false)" and some use "NotNullWhen(true)" -- do these mean different things? Why are some one and some the other?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs [hunk 17]
```diff
@@ -273,6 +274,33 @@ public TimeSpan Expect100ContinueTimeout
             }
         }
 
+        /// <summary>
+        /// When non-null, a custom factory used to open new TCP connections.
+        /// When null, a <see cref="SocketsHttpConnectionFactory"/> will be used.
+        /// </summary>
+        public ConnectionFactory? ConnectionFactory
```
Comment (geoffkizer) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:281]: Why is this nullable? Why don't we just return the default factory instead of null?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 18]
```diff
@@ -1096,73 +1108,101 @@ private async ValueTask<(Socket?, Stream?, TransportContext?, HttpResponseMessag
...
-                return (socket, stream, transportContext, null);
+                return (connection, transportContext, null);
             }
             finally
             {
                 cancellationWithConnectTimeout?.Dispose();
             }
         }
 
+        private ValueTask<Connection> ConnectToTcpHostAsync(string host, int port, HttpRequestMessage initialRequest, bool async, CancellationToken cancellationToken)
```
Comment (geoffkizer) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1156]: This seems like a weird name... or perhaps I'm missing something here. The connection factory can return literally anything, not necessarily a TCP connection, right? So why is this called ConnectToTcpHost?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 19]
```diff
@@ -1096,73 +1108,101 @@ private async ValueTask<(Socket?, Stream?, TransportContext?, HttpResponseMessag
...
+            catch (Exception ex)
+            {
+                return ValueTask.FromException<Connection>(ex);
+            }
+        }
+
         internal async ValueTask<(HttpConnection?, HttpResponseMessage?)> CreateHttp11ConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
         {
-            (Socket? socket, Stream? stream, TransportContext? transportContext, HttpResponseMessage? failureResponse) =
+            (Connection? connection, TransportContext? transportContext, HttpResponseMessage? failureResponse) =
```
Comment (geoffkizer) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1186]: With all this new connection related infrastructure, I have to wonder why `transportContext` isn't just a connection property or something like that.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Connections/TaskSocketAsyncEventArgs.cs [hunk 20]
```diff
@@ -0,0 +1,27 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Net.Sockets;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Sources;
+
+namespace System.Net.Connections
+{
+    internal sealed class TaskSocketAsyncEventArgs : SocketAsyncEventArgs, IValueTaskSource
```
Comment (geoffkizer) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Connections/TaskSocketAsyncEventArgs.cs:10]: Is there really any benefit to this anymore? Why can't we just use Task-based async methods on Socket directly?

If there is some benefit, it would be nice to understand what it is, and fix the Task-based methods if possible.

**Review by geoffkizer — APPROVED** [review]
A few questions and minor feedback above, but in general looks fine.

File: src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs]: It is shadowing because these are nested types. Nested types have access to private fields/methods of the parent.

If you moved the private types out of the parent type then it would no longer be shadowing.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs (hunk 18 above)
Comment (scalablecory) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1156]: `ConnectToWhatWouldNormallyBeATcpHost` was too long.

File: src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs (hunk 16 above)
Comment (scalablecory) [reply src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs:23]: `NotNullWhen(true)` does not work when T is non-nullable. @stephentoub recommended `MaybeNullWhen(false)` as a workaround that works with non-nullable types.

I don't 100% understand the intricacies of this feature yet myself. I can't really tell you what the difference between the two is, other than one works here and the other doesn't. @buyaa-n can you help?

Comment (stephentoub) [reply src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs:23]: See the second bullet under https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/api-guidelines/nullability.md#nullable-attributes

File: src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs (hunk 15 above)
Comment (scalablecory) [reply src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs]: Good catch.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs (hunk 17 above)
Comment (scalablecory) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:281]: IIRC no strong reasons, it's just the way we decided to go.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs (hunk 19 above)
Comment (scalablecory) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1186]: I'm pretty sure it's just taken from the `SslStream`, so this change would be trivial.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Connections/TaskSocketAsyncEventArgs.cs (hunk 20 above)
Comment (scalablecory) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Connections/TaskSocketAsyncEventArgs.cs:10]: This is used to get access at some features not yet exposed via TAP methods -- like cancellable connect and gathered writes.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs (hunk 19 above)
Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1186]: It is, yeah.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Connections/TaskSocketAsyncEventArgs.cs (hunk 20 above)
Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Connections/TaskSocketAsyncEventArgs.cs:10]: It's super lame that we don't expose cancellable connect via TAP. Is there an issue on this?

Gathered write would be good to do as well, but it's a bit more complicated so it's understandable why we don't have it currently.

Comment (scalablecory) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Connections/TaskSocketAsyncEventArgs.cs:10]: https://github.com/dotnet/runtime/issues/33418 and https://github.com/dotnet/runtime/issues/33417

File: src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs (hunk 16 above)
Comment (buyaa-n) [reply src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs:23]: in short `NotNullWhen(bool)` used for reference types, MaybeNullWhen(bool) used for generic types where it can be a value type where it would never be null

**Comment by davidfowl** [comment]
> System.IO.Pipelines is now included in Net50 SDK.

MVP ✌🏾

File: src/libraries/System.Net.Connections/Directory.Build.props [hunk 21]
```diff
@@ -2,6 +2,5 @@
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
     <StrongNameKeyId>Microsoft</StrongNameKeyId>
-    <IsNETCoreApp>true</IsNETCoreApp>
```
Comment (ViktorHofer) [inline src/libraries/System.Net.Connections/Directory.Build.props]: These need to be specified here now: https://github.com/dotnet/runtime/blob/master/src/libraries/NetCoreAppLibrary.props#L4

File: src/libraries/System.IO.Pipelines/Directory.Build.props [hunk 22]
```diff
@@ -2,6 +2,5 @@
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
     <StrongNameKeyId>Open</StrongNameKeyId>
-    <IsNETCoreApp>true</IsNETCoreApp>
```
Comment (ViktorHofer) [inline src/libraries/System.IO.Pipelines/Directory.Build.props]: Same for this one: https://github.com/dotnet/runtime/blob/master/src/libraries/NetCoreAppLibrary.props#L4

File: src/libraries/System.Net.Connections/Directory.Build.props (hunk 21 above)
Comment (scalablecory) [reply src/libraries/System.Net.Connections/Directory.Build.props]: Thanks!

**Comment by omariom** [comment]
@scalablecory   Will it be available as a .NET Standard nuget?

**Comment by karelz** [comment]
@omariom it is tied to `SocketsHttpHandler` which does not seem to be in .NET Standard ...
Also note that some platforms might have `SocketsHttpHandler` available, but not as primary/recommended handler -- e.g. Android, iOS/tvOS/*OS

That said, I am not sure what is the exact goal/value of .NET Standard these days ...

**Comment by stephentoub** [comment]
> it is tied to SocketsHttpHandler

@karelz, it isn't.  SocketsHttpHandler is one consumer of the APIs.

**Comment by karelz** [comment]
OK, sorry for the wrong answer then, I'll hide my reply to avoid further confusion and let others answer ...

**Comment by scalablecory** [comment]
> @scalablecory Will it be available as a .NET Standard nuget?

If you have use cases for this as a .NET Standard package, please open a new issue so we can gauge interest and discuss further.

No concrete plans to announce here but we have been discussing this. Specifically, @davidfowl indicated interest to use this in the .NET Standard SignalR client library.

Note - even if these core abstractions make it into a package, HttpClient's support for it will still only be in .NET 5.

File: src/libraries/System.IO.Pipelines/pkg/System.IO.Pipelines.pkgproj [hunk 23]
```diff
@@ -11,6 +11,7 @@
     </ProjectReference>
     <ProjectReference Include="..\src\System.IO.Pipelines.csproj" />
     <HarvestIncludePaths Include="lib/netstandard1.3" />
+    <InboxOnTargetFramework Include="net5.0" />
```
Comment (ericstj) [inline src/libraries/System.IO.Pipelines/pkg/System.IO.Pipelines.pkgproj:14]: This is bad.  Don't do this.  It freezes the API.  Either stop producing the package, or treat it as OOB if you still want to control the API like an OOB (EG: System.Text.Json, ImmutableCollections, etc.)

Comment (davidfowl) [reply src/libraries/System.IO.Pipelines/pkg/System.IO.Pipelines.pkgproj:14]: We should be doing the latter. It's just like System.Text.Json etc.

Comment (scalablecory) [reply src/libraries/System.IO.Pipelines/pkg/System.IO.Pipelines.pkgproj:14]: CC @geoffkizer

Comment (ericstj) [reply src/libraries/System.IO.Pipelines/pkg/System.IO.Pipelines.pkgproj:14]: I went ahead and submitted a fix for this: https://github.com/dotnet/runtime/pull/40212

File: src/libraries/System.Net.Connections/src/System/Net/Connections/DuplexPipeStream.cs [hunk 24]
```diff
@@ -0,0 +1,166 @@
...
+
+            ValueTask<int> t = ReadAsync(buffer.AsMemory(offset, count));
+            return
+                t.IsCompleted ? t.GetAwaiter().GetResult() :
+                t.AsTask().GetAwaiter().GetResult();
+        }
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null) return Task.FromException<int>(ExceptionDispatchInfo.SetCurrentStackTrace(new ArgumentNullException(nameof(buffer))));
```
Comment (stephentoub) [inline src/libraries/System.Net.Connections/src/System/Net/Connections/DuplexPipeStream.cs:70]: Argument exceptions should just be thrown rather than stored into the returned task.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 25]
```diff
@@ -1223,7 +1263,12 @@ private async ValueTask<(Stream?, HttpResponseMessage?)> EstablishProxyTunnel(bo
                 return (null, tunnelResponse);
             }
 
-            return (tunnelResponse.Content!.ReadAsStream(cancellationToken), null);
+            Stream stream = tunnelResponse.Content.ReadAsStream(cancellationToken);
+            EndPoint remoteEndPoint = new DnsEndPoint(_originAuthority.IdnHost, _originAuthority.Port);
+
+            // TODO: the Socket from the response can be funneled into a connection property here.
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1269]: Is there an issue tracking this?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpConnectionFactory.cs [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+    public class SocketsHttpConnectionFactory : ConnectionFactory
+    {
+        internal static SocketsHttpConnectionFactory Default { get; } = new SocketsHttpConnectionFactory();
+
+        /// <inheritdoc/>
+        public sealed override ValueTask<Connection> ConnectAsync(EndPoint? endPoint, IConnectionProperties? options = null, CancellationToken cancellationToken = default)
+        {
+            if (options == null || !options.TryGet(out DnsEndPointWithProperties? httpOptions))
+            {
+                return ValueTask.FromException<Connection>(ExceptionDispatchInfo.SetCurrentStackTrace(new HttpRequestException($"{nameof(SocketsHttpConnectionFactory)} requires a {nameof(DnsEndPointWithProperties)} property.")));
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpConnectionFactory.cs:24]: SetCurrentStackTrace is really only useful when we expect the call chain and the await chain to be different, e.g. if an exception occurs, is stored into a field, and is then later thrown from that field.  Here, presumably we expect the 99% case to be `await ConnectAsync(...)`, in which case the SetCurrentStackTrace here is not only unnecessary.
